cache/demux.go

```diff
@@ -0,0 +1,146 @@
+package cache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+type demux struct {
+	mu              sync.RWMutex
+	activeWatchers  map[*watcher]struct{}
+	laggingWatchers map[*watcher]struct{}
+	history         History
+	resyncInterval  time.Duration
+}
+
+// newDemux launches a background resyncLoop that stops when ctx is cancelled.
+func newDemux(ctx context.Context, wg *sync.WaitGroup, history History, resyncInterval time.Duration) *demux {
+	d := &demux{
+		activeWatchers:  make(map[*watcher]struct{}),
+		laggingWatchers: make(map[*watcher]struct{}),
+		history:         history,
+		resyncInterval:  resyncInterval,
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		d.resyncLoop(ctx)
+	}()
+	return d
+}
+
+func (d *demux) add(w *watcher) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.activeWatchers[w] = struct{}{}
+}
+
+func (d *demux) remove(w *watcher) {
+	func() {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		delete(d.activeWatchers, w)
+		delete(d.laggingWatchers, w)
+	}()
+	w.Stop()
+}
+
+func (d *demux) broadcast(event *clientv3.Event) {
+	d.mu.RLock()
+	active := make([]*watcher, 0, len(d.activeWatchers))
```

Comment:
Don't copy, just execute everything under lock. Let's not do premature optimizations as they can make code harder to understand and prone to subtle concurrency issues.